### PR TITLE
feat(ui): Add `highlightedRowKey` prop to `GridEditable`

### PR DESCRIPTION
**Pull request by gggritso**
Adds a way to specify which row in a table should be "highlighted". A "highlighted row" shows up a little differently in the UI.

We have functionality in Performance where hovering on a sample plotted in a chart will highlight the corresponding row in the table, so people know which row in the table corresponds with which sample. 

Right now, we accomplish that by wrapping every data cell in a special `div` and setting styles on it. This is very awkward, since it makes it harder to use field renderers, and requires some yucky setup work.

Here, I introduce a prop to allow showing a highlighted row in a simpler way.

**e.g.,**
Screenshot 2024-04-04 at 4 52 49 PM

## Questions

- what should a highlighted row look like? In Performance, we _bold_ it. Here, I decided to add a background colour (arbitrarily chosen). What's best?
- is this API sensible? I waffled on this, since I thought it might be better to provide a matcher function, but this was the simplest approach by a _lot_

### CODE REVIEW

**Comment by gggritso** [comment]
@vuluongj20 would love your thoughts on this!

**Comment by evanpurkhiser** [comment]
Tagging Vu in on this!

File: static/app/components/gridEditable/index.stories.tsx [hunk 1]
```diff
@@ -141,12 +144,13 @@ export default storyBook(GridEditable, story => {
...
-          onRowMouseOver={dataRow => {
-            setActiveRow(dataRow);
+          onRowMouseOver={(_dataRow, key) => {
+            setActiveRowKey(key);
           }}
           onRowMouseOut={() => {
-            setActiveRow(undefined);
+            setActiveRowKey(undefined);
           }}
+          highlightedRowKey={activeRowKey}
```
Comment (ryan953) [inline static/app/components/gridEditable/index.stories.tsx:153]: I dig this api 👍

**Comment by vuluongj20** [comment]
Lol what a co-incidence I'm working on a Table refactor proposal right now:
image

In terms of implementation, should we perhaps store the hovered row index state inside the component? That state can be overridden via a prop, but otherwise state management is automatic so the parent component doesn't have to store that state. Wdyt?

**Comment by gggritso** [comment]
@vuluongj20 That makes sense to me! I can put that in, and we can see how it's looking. What's the correct background colour?

**Comment by gggritso** [comment]
Bad news. In short, storing the highlighted index in the component state doesn't work well. Each time the state is updated, the table re-renders, which takes _200ms_ in some tables (e.g., Performance) even on my computer.

It's tough to memoize table cells, since they depend only on `children` which are generated semi-from-scratch, but maybe it's possible? I went with a manual DOM approach, which feels bad but performs well.

@vuluongj20 what do you think?

**Comment by vuluongj20** [comment]
> Bad news. In short, storing the highlighted index in the component state doesn't work well. Each time the state is updated, the table re-renders, which takes _200ms_ in some tables (e.g., Performance) even on my computer.
> 
> It's tough to memoize table cells, since they depend only on `children` which are generated semi-from-scratch, but maybe it's possible? I went with a manual DOM approach, which feels bad but performs well.
> 
> @vuluongj20 what do you think?

@gggritso Hmm I think this is worth further investigation (I'm not exactly sure why the re-render happens on every row since we only call `highlightRow`, there might be a way to prevent it without memoization), but for now we can start with an external prop. 

As for the background, is there a way we can use [`InteractionStateLayer`](https://github.com/getsentry/sentry/blob/master/static/app/components/interactionStateLayer.tsx#L14) here? That's the best way to do hover effects in our design system.

**Comment by gggritso** [comment]
@vuluongj20 that works for me 👌🏻 Reverted back to using an external prop, and switched to `InteractionStateLayer`. Happy to look at this again in the future if you need me

**Comment by gggritso** [comment]
@vuluongj20 actually I take it back, Jest is furious because it's not valid to place an `InteractionStateLayer` inside a `tr` because the layer is a `span` and that's invalid DOM nesting. Is it safe to change the interaction layer to a `div` since it's basically acting as a block element? If not, maybe I can make the DOM element type configurable 🤔

**Review by vuluongj20 — APPROVED** [review]
Just tested and it looks awesome! Love the fact that the pressed state also works!

**Pull request merged**